xdsl/dialects/builtin.py

```diff
@@ -1466,9 +1466,10 @@ def constr(
     ) -> AttrConstraint[TensorType[AttributeInvT]]:
         if element_type is None and shape is None:
             return BaseAttr[TensorType[AttributeInvT]](TensorType)
+        element_type_constr = AnyAttr() if element_type is None else element_type
         shape_constr = AnyAttr() if shape is None else shape
         return ParamAttrConstraint[TensorType[AttributeInvT]](
-            TensorType, (shape_constr, element_type, AnyAttr())
+            TensorType, (shape_constr, element_type_constr, AnyAttr())
```

Comment:
I'm surprised this is needed, what happens if you run the test with this change reverted?

Reply:
Actually, it just works. I guess this change isn't needed. Should I close the PR?

Reply:
Let's merge the test? It'll be a kind of documentation for the constraint also.